arraykit.c

```diff
@@ -132,6 +132,87 @@ name_filter(PyObject *Py_UNUSED(m), PyObject *n)
     return n;
 }
 
+
+static PyObject *
+shape_filter(PyObject *Py_UNUSED(m), PyObject *a)
+{
+    // If array is ndim 1, return (shape[0], 1), else return shape.
+    // https://numpy.org/doc/stable/reference/c-api/array.html
+    // https://stackoverflow.com/questions/56182259/how-does-one-acces-numpy-multidimensionnal-array-in-c-extensions
+
+    PyObject *shape = PyTuple_New(2);
+    if (!shape) {
+        return NULL;
+    }
+
+    PyArrayObject *array;
+    array = (PyArrayObject *)a;
+
+    // get size at dimension 0, set in tuple position 0
+    int size0 = PyArray_DIM(array, 0);
+    PyObject *rows = PyLong_FromLong(size0);
+    PyTuple_SET_ITEM(shape, 0, rows);
+
+    // handle 1 dimensional array
+    if (PyArray_NDIM(array) == 1 ) {
+        PyObject *cols = PyLong_FromLong(1);
+        PyTuple_SET_ITEM(shape, 1, cols);
+        return shape;
+    }
+
+    // get size at dimension 1, set in tuple position 1
+    int size1 = PyArray_DIM(array, 1);
+    PyObject *cols = PyLong_FromLong(size1);
+    PyTuple_SET_ITEM(shape, 1, cols);
+    return shape;
+
+}
+
+
+static PyObject *
+column_2d_filter(PyObject *Py_UNUSED(m), PyObject *a)
+{
+    // If array ndim is 1, reshape into ndim 2 with 1 column
+    // related example: https://github.com/RhysU/ar/blob/master/ar-python.cpp
+
+    PyArrayObject *array;
+    array = (PyArrayObject *)a;
+
+    if (PyArray_NDIM(array) == 1 ) {
+        // https://numpy.org/doc/stable/reference/c-api/types-and-structures.html#c.PyArray_Dims
+        npy_intp dim[2] = {PyArray_DIM(array, 0), 1};
+        PyArray_Dims shape = {dim, sizeof(dim)/sizeof(dim[0])};
+
+        PyObject *array_new;
+        array_new = PyArray_Newshape(array, &shape, NPY_ANYORDER);
+
+        if (!array_new) {
+            Py_DECREF(array_new);
+            PyErr_SetString(PyExc_RuntimeError,  "Unable to reshape array.");
+            return NULL;
+        }
+
+        return array_new; // already a PyObject*
```

Comment:
`PyArray_Newshape` should have already set a helpful error on failure. If you want to raise something different, you need to clear the current one first with `PyErr_Clear()` (but I don't recommend it here).

So since it either returns a new array or returns `NULL` with an error set, there's nothing left to do!

```suggestion
        return PyArray_Newshape(array, &shape, NPY_ANYORDER);
```

Reply:
Thanks for explaining this!